### PR TITLE
[processing] Make GDAL/OGR algorithms work with (input) memory layers

### PR DESCRIPTION
**Pull request by gacarrillor**
In general, all input vectors in GDAL/OGR algorithms will be exported to GeoPackage preferably (**only** when it's needed, e.g., when we use memory layers, when the alg is going to use selected features and when special characters are found in layer.source()). The rationale behind the support of GeoPackage as export format is to avoid problems such as field name truncation that may occur exporting to Shapefiles.

Notes related to GeoPackage and OGR:  
1. If you want to pass a SQL clause to OGR for creating a GeoPackage file, the wild card `*` (e.g., `SELECT ST_Buffer(geom,20),* FROM table`) throws an error related to "duplicate column name: geom". `geom` is the default geometry column name for GeoPackage and from [the docs](http://www.gdal.org/drv_geopackage.html), "GeoPackage only supports one geometry column per table."
2. The SQL wild card `*` is (strangely) more tolerant when geometry column is called 'geometry'; if it's called 'geom', `*` fails. This led me to use `geometry` as default geometry column name for temporary GPKG files. Nonetheless, I've replaced `*` in OGR algorithms for a complete list of input layer fields, so that Processing can deal with both geometry column names: 'geom' and 'geometry'.
3. There seems to be an issue with precision in GeoPackage support, since the extent of a GeoPackage file turned out to be different from the source layer's extent after running `Convert format` alg. See [this line in the test](https://github.com/gacarrillor/QGIS/blob/a375f6fcd7665cdbdd50e4f014c40cc1e66f3bb9/python/plugins/processing/tests/GdalAlgorithmsMemoryTest.py#L109). I keep getting `270779.0,4458920.0 : 270862.0,4458980.0` vs. `270779.2129507259815000,4458920.3218659535050392 : 270861.6575738972751424,4458980.6503353714942932`

Other additions/fixes:
- Processing now passes the `-f` parameter to OGR algorithms depending on the extension of the output parameter. This parameter was not being passed by most algorithms, so only Shapefiles were being generated.
- Processing now passes the `-a_srs` parameter to OGR commands based on the input layer's CRS. This parameter was not being passed by most algorithms, so formats like GeoJSON couldn't be loaded properly to QGIS due to missing CRS.
- Now `tools.dataobjects.exportVectorLayer` does not export a layer if its extension is in supported list. It was comparing things like ".geojson" vs. "geojson", therefore always exporting. 
- Removing "." from base names to avoid issues with table names in Spatialite.

When referring to a memory layer in "Creation options" parameters, the user needs to use `memory_layer` as layer name, for example:

```
     -dialect sqlite -sql "SELECT ST_Buffer( geometry , 20 ), * FROM 'memory_layer' "
     -dialect sqlite -sql 'SELECT ST_Buffer( geometry , 20 ), * FROM "memory_layer" '
     -dialect sqlite -sql "SELECT ST_Buffer( geometry , 20 ), * FROM \"memory_layer\" "
```

Documentation should recommend using double quotes for identifiers (when needed by column and table names) and single quotes for values. (See http://trac.osgeo.org/gdal/wiki/rfc52_strict_sql_quoting) By and large, It's safer to use double quotes for the memory_layer table name (i.e., second and third options above), as well as for table names corresponding to other formats.

Includes tests for 
- vector.ogrLayerName, 
- dataobjects.exportVectorLayer,
- memory layers as input of 3 OGR algorithms. 

[needs-docs]
Fix #14430

### CODE REVIEW

**Comment by palmerj** [comment]
I would suggest making geopackage the standard output format because it based on a standard, where spatialite is not.

**Comment by m-kuhn** [comment]
> In preliminary tests, temporary databases are about 5.8MB, which seems a lot (since corresponding test Shapefiles were about 2kB). One could think of reusing Spatialite DBs per processing session or so, but I'll leave that for another person willing to contribute :).

That shouldn't be the case with GeoPackage.

**Comment by gacarrillor** [comment]
Thanks to both. Indeed, GeoPackage adds no generation overhead and temporary files are way smaller. I've updated the PR description, please re-read it.

**Comment by palmerj** [comment]
> Notes related to GeoPackage and OGR:

I would say @rouault would be best to comment of these issues.

**Comment by m-kuhn** [comment]
I guess you have a self-compiled QGIS?
In this case you can just run the tests from the build directory

    ctest -R -V ProcessingGdalAlgorithmsMemoryTest

File: python/plugins/processing/tests/GdalAlgorithmsMemoryTest.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+from processing.core.ProcessingConfig import ProcessingConfig
+from PyQt4.QtCore import QVariant
+
+import os.path
+import errno
+import shutil
+
+dataFolder = os.path.join(os.path.dirname(__file__), 'testdata/')
+expectedFolder = os.path.abspath(os.path.join(dataFolder, 'expected', 'gdal'))
+tmpBaseFolder = os.path.join(os.sep, 'tmp', 'qgis_test', str(os.getpid()))
```
Comment (m-kuhn) [inline python/plugins/processing/tests/GdalAlgorithmsMemoryTest.py]: Better use a portable approach like the [one here](https://github.com/gacarrillor/QGIS/blob/a375f6fcd7665cdbdd50e4f014c40cc1e66f3bb9/python/plugins/processing/tests/AlgorithmsTestBase.py#L154-L158)

File: python/plugins/processing/tests/GdalAlgorithmsMemoryTest.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                       QgsDataSourceURI, QgsField)
+from processing.core.ProcessingConfig import ProcessingConfig
+from PyQt4.QtCore import QVariant
+
+import os.path
+import errno
+import shutil
+
+dataFolder = os.path.join(os.path.dirname(__file__), 'testdata/')
+expectedFolder = os.path.abspath(os.path.join(dataFolder, 'expected', 'gdal'))
```
Comment (m-kuhn) [inline python/plugins/processing/tests/GdalAlgorithmsMemoryTest.py]: Works also without abspath

File: python/plugins/processing/tests/GdalAlgorithmsMemoryTest.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+tmpBaseFolder = os.path.join(os.sep, 'tmp', 'qgis_test', str(os.getpid()))
+
+
+def mkDirP(path):
```
Comment (m-kuhn) [inline python/plugins/processing/tests/GdalAlgorithmsMemoryTest.py]: Not required with the `tempfile` approach

File: python/plugins/processing/algs/gdal/ogr2ogrdissolve.py [hunk 4]
```diff
@@ -95,8 +99,9 @@ def getConsoleCommands(self):
         fields = self.getParameterValue(self.FIELDS)
         querystart = '-dialect sqlite -sql "SELECT ST_Union(' + geometry + ')'
         queryend = ' FROM ' + layername + ' GROUP BY ' + field + '"'
+        fieldList = getLayerFieldList(inLayer)
         if fields:
-            queryfields = ",*"
+            queryfields = ',%s' % fieldList
```
Comment (m-kuhn) [inline python/plugins/processing/algs/gdal/ogr2ogrdissolve.py]: Sidenote, it's advisable to switch to the newer style syntax `',{}'.format(fieldList)` ;)

**Comment by m-kuhn** [comment]
Thanks @gacarrillor , looks like clean code and some nice refactorings.

**Comment by luipir** [comment]
as far as I remember there is a gdal/GeoPackage fix for the extent problem... this fix will be incorporated in the current gdal RC. btw no impediments to use memory layer.

File: python/plugins/processing/tools/vector.py [hunk 5]
```diff
@@ -580,6 +591,103 @@ def ogrLayerName(uri):
     return name
 
 
+# If you add something to ogrDisplayFormats, make sure you also update
+# ogrFormats and ogrExtensions. The order matters!
```
Comment (gacarrillor) [inline python/plugins/processing/tools/vector.py]: I'll let this one for a follow-up PR.

**Comment by gacarrillor** [comment]
Thanks for the review Matthias. Concerning the `ctest` command, I keep getting a `No tests were found!!!` message. 

Unfortunately, this PR couldn't make it into QGIS v2.18. I don't know if it can be included (I mean, easily) into 2.18.x. Nonetheless, I'll let this PR here as a basis for an (hopefully) upcoming PR for v3.0.

**Comment by jef-n** [comment]
Automatically closed on removal of the master_2 branch.

**Comment by m-kuhn** [comment]
> Thanks for the review Matthias. Concerning the ctest command, I keep getting a No tests were found!!! message.

Did you cmake with `-DENABLE_TESTS`?

**Comment by gacarrillor** [comment]
> Did you cmake with -DENABLE_TESTS?

Yes, tests were enabled in the `ccmake ..` configuration. 

To run a particular test I use this (seen in the qgis-dev mailing list):

```
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/apps/lib/
QGIS_PREFIX_PATH=/apps PYTHONPATH=/apps/share/qgis/python:/apps/share/qgis/python/plugins:$PYTHONPATH python ../python/plugins/processing/tests/ParametersTest.py
```

However, as you can see in Travis, my test is still failing and still getting a segmentation fault. I'll try with QGIS v3 in the next few days.

**Pull request closed without merging**